### PR TITLE
Adding a Redis tutorial

### PR DESCRIPTION
**Pull request by therealbill**
This is a touch more than a basic Redis setup in that it addresses some minimal security options w/Redis.

### CODE REVIEW

**Comment by lauracly** [comment]
Two questions - I got to step 2 and my container wasn't up. The log command didn't work for me - kept asking for 1 required argument, but no matter what I added I didn't get any logs. So I kept going, and when I got to step 4 I got an another error "No public port '6379/tcp' published for <container>". I am in the office but running on the VPN (I turned that off and still got the same thing). Should I try it at home completely disconnected from the work network?

**Comment by lauracly** [comment]
OK got past those issues working at home. Now I run into this error: bash: redis-cli: command not found. 
The container is up and running, and everything else is working.

Do I need to do something to get the cli to work?

**Comment by everett-toews** [comment]
Looks like there's an assumption that the user has redis-cli installed. Options are:

(a) Add redis-cli to the Prerequisites
or
(b) Run redis-cli in another container and connect to the host from there.

My preference is (b). The user won't have to install anything locally and it demonstrates connecting to the Redis host from another container.

**Comment by therealbill** [comment]
OK I'll add running it in a container.

**Comment by therealbill** [comment]
@lauracly yeah connecting from the corp net (or VPN) is problematic. I usually run a Cloud Server to do the work/testing from to avoid it.

File: _tutorials/data-stores-redis.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    $ echo $REDIS_HOST:$REDIS_PORT
+    172.99.78.231:32769
+    ```
+
+	Now let us test our connection. To do this we will run redis-cli in it's own container as well. 
+	For this we will still use the official Redis Docker image, we will simply be passing a 
+	different command to it.
+
+    ```bash
+	$ run --interactive --tty  redis redis-cli  -h $REDIS_HOST -p $REDIS_PORT -a mysecretpassword
```
Comment (everett-toews) [inline _tutorials/data-stores-redis.md]: Add `--rm` to remove the container automatically when it exits.

File: _tutorials/data-stores-redis.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+	Now type in `PING` and you should see `PONG` come back. You should
+	now have a functional Redis server requiring authentication. Now
+	we'll set and get a key. This we'll do using the the container call 
+	directly. First, exit the existing redis-cli session by running the 
+	`quit` command. You should no be back to your regular command line.
+	
+	Next up, let us set the key `redis:on:carina` to a value of "true".
+	```bash
+	$ run --interactive --tty  redis redis-cli  -h $REDIS_HOST \
```
Comment (everett-toews) [inline _tutorials/data-stores-redis.md]: Add `--rm` to remove the container automatically when it exits.

File: _tutorials/data-stores-redis.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+	  -p $REDIS_PORT -a mysecretpassword  set redis:on:carina true
+	OK
+	```
+	
+	Now, `get` the key's value to see it was set.
+	```bash
+	$ run --interactive --tty  redis redis-cli  -h $REDIS_HOST \
```
Comment (everett-toews) [inline _tutorials/data-stores-redis.md]: Add `--rm` to remove the container automatically when it exits.

File: _tutorials/data-stores-redis.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	`quit` command. You should no be back to your regular command line.
+	
+	Next up, let us set the key `redis:on:carina` to a value of "true".
+	```bash
+	$ run --interactive --tty  redis redis-cli  -h $REDIS_HOST \
+	  -p $REDIS_PORT -a mysecretpassword  set redis:on:carina true
+	OK
+	```
+	
+	Now, `get` the key's value to see it was set.
```
Comment (everett-toews) [inline _tutorials/data-stores-redis.md]: Needs newline here

File: _tutorials/data-stores-redis.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+	```bash
+	$ run --interactive --tty  redis redis-cli  -h $REDIS_HOST \
+	  -p $REDIS_PORT -a mysecretpassword  get redis:on:carina
+	"true"
+	```
```
Comment (everett-toews) [inline _tutorials/data-stores-redis.md]: Needs newline here

File: _tutorials/data-stores-redis.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+	REDIS_PORT for the port.
+
+    ```bash
+    $ REDIS_HOST=$(docker inspect --format '{{ (index (index .NetworkSettings.Ports "6379/tcp") 0).HostIp }}' $REDCON)
+    $ REDIS_PORT=$(docker inspect --format '{{ (index (index .NetworkSettings.Ports "6379/tcp") 0).HostPort }}' $REDCON)
+    $ echo $REDIS_HOST:$REDIS_PORT
+    172.99.78.231:32769
+    ```
+
+	Now let us test our connection. To do this we will run redis-cli in it's own container as well.
```
Comment (everett-toews) [inline _tutorials/data-stores-redis.md]: This should be the beginning of another step (i.e. add a `1.` as the above steps).

File: _tutorials/data-stores-redis.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    ```
+
+	This `docker run` command will output the ID of your running Redis container.
+	The `-m 1G` option specifies a reserved memory of 1GB. With Redis all data
+	is in memory so you'll want to reserve some meory for the contaier and
+	folow it up with a `maxmemory` setting to limit Redis' data memory
+	allocation. In this case we selected 800Mb as the maxmemory value for Redis
+	to provide 800M of data+client buffer space and 200M for Redis persistence
+	requirements.
+
```
Comment (everett-toews) [inline _tutorials/data-stores-redis.md]: Please make the indenting consistent throughout the whole tutorial. Anything within the body of a step should be indented 4 spaces, whether it's a code block or prose.

**Comment by everett-toews** [comment]
This is looking good! Just a few last minor quibbles.

File: _tutorials/data-stores-redis.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+See [Troubleshooting common problems](/docs/tutorials/troubleshooting/).
+
+For additional assistance, ask the [community](https://community.getcarina.com/) for help or join us in IRC at [#carina on Freenode](http://webchat.freenode.net/?channels=carina).
+
+### Resources
+
+* [The Redis Quickstart](http://redis.io/topics/quickstart)
+* [The Redis Security Guide](http://redis.io/topics/security)
+
+### Next
```
Comment (everett-toews) [inline _tutorials/data-stores-redis.md]: `### Next steps`

File: _tutorials/data-stores-redis.md [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+If Redis isn't the data store for you, read [Use MongoDB on Carina](/docs/tutorials/data-stores-mongodb/).
```
Comment (everett-toews) [inline _tutorials/data-stores-redis.md]: Add another line.

`If you want Redis to only listen on the internal, multi-tenant network within Carina, read [Communicate between containers over the ServiceNet internal network](/docs/tutorials/servicenet/).`

**Comment by rgbkrk** [comment]
@therealbill what do you think about @everett-toews's suggestions?

**Comment by therealbill** [comment]
@rgbkrk I think they're good. I've been on vacation so I've been offline for a while. ;)

**Comment by rgbkrk** [comment]
Understandably. Just let us know when you're ready to go!

File: _tutorials/data-stores-redis.md (hunk 1 above)
Comment (everett-toews) [inline _tutorials/data-stores-redis.md]: Missing `docker` before `run`

File: _tutorials/data-stores-redis.md (hunk 2 above)
Comment (everett-toews) [inline _tutorials/data-stores-redis.md]: Missing `docker` before `run`

File: _tutorials/data-stores-redis.md (hunk 3 above)
Comment (everett-toews) [inline _tutorials/data-stores-redis.md]: Missing `docker` before `run`

**Comment by therealbill** [comment]
OK, back from vacation and I believe I've integrated all of the suggestions and corrections. Please let me know if I've missed any.

**Comment by lauracly** [comment]
Looks good to me. @everett-toews you agree?

**Comment by rgbkrk** [comment]
Update the date for today or tomorrow, then I say :shipit:!

**Comment by therealbill** [comment]
So, good to go, or wait for @everett-toews to +1 it?

**Comment by lauracly** [comment]
Hey - Everett and I are good. Just talked to Kelly. She will review and send you any final edits via a Word doc and post it here. Once those are done, then yes we can merge.

**Comment by therealbill** [comment]
@lauracly thanks

**Comment by everett-toews** [comment]
:+1: from me after the edits from @kmsholcomb are done.

**Comment by kmsholcomb** [comment]
Hi @therealbill: My suggested copy edits are in the attached Word document. I used Track Changes to suggest additions, deletions, and changes, and Commenting to ask some questions. Please let me know if you have any questions about interpreting the markup or about the edits themselves. Thanks!

[Use Redis on Carina_editsKH.docx](https://github.com/getcarina/getcarina.com/files/78827/Use.Redis.on.Carina_editsKH.docx)

**Comment by rgbkrk** [comment]
Looks like the edits went in, merging.

**Pull request merged**